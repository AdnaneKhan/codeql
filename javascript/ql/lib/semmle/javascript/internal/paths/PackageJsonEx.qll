private import javascript
private import semmle.javascript.internal.paths.PathResolver
private import semmle.javascript.internal.paths.PathExprResolver

class PackageJsonEx extends PackageJson {
  private JsonValue getAPartOfExportsSection(string pattern) {
    result = this.getPropValue("exports") and
    pattern = ""
    or
    exists(string prop, string prevPath |
      result = this.getAPartOfExportsSection(prevPath).getPropValue(prop) and
      if prop.matches("./%") then pattern = prop.suffix(2) else pattern = prevPath
    )
  }

  predicate hasPathMapping(string pattern, string newPath) {
    this.getAPartOfExportsSection(pattern).getStringValue() = newPath
  }

  predicate hasExactPathMapping(string pattern, string newPath) {
    this.getAPartOfExportsSection(pattern).getStringValue() = newPath and
    not pattern.matches("%*%")
  }

  predicate hasPrefixPathMapping(string pattern, string newPath) {
    this.hasPathMapping(pattern + "*", newPath + "*")
  }

  predicate hasExactPathMappingTo(string pattern, Container target) {
    exists(string newPath |
      this.hasExactPathMapping(pattern, newPath) and
      target = Resolver::resolve(this.getFolder(), newPath)
    )
  }

  predicate hasPrefixPathMappingTo(string pattern, Container target) {
    exists(string newPath |
      this.hasPrefixPathMapping(pattern, newPath) and
      target = Resolver::resolve(this.getFolder(), newPath)
    )
  }

  string getMainPath() { result = this.getPropStringValue(["main", "module"]) }

  Container getMainFile() { result = Resolver::resolve(this.getFolder(), this.getMainPath()) }

  string getAPathInFilesArray() {
    result = this.getPropValue("files").(JsonArray).getElementStringValue(_)
  }

  Container getAFileInFilesArray() {
    result = Resolver::resolve(this.getFolder(), this.getAPathInFilesArray())
  }
}

private module ResolverConfig implements PathResolverSig {
  additional predicate shouldResolve(PackageJsonEx pkg, Container base, string path) {
    base = pkg.getJsonFile().getParentContainer() and
    (
      pkg.hasExactPathMapping(_, path)
      or
      pkg.hasPrefixPathMapping(_, path)
      or
      path = pkg.getMainPath()
      or
      path = pkg.getAPathInFilesArray()
    )
  }

  predicate shouldResolve(Container base, string path) { shouldResolve(_, base, path) }

  predicate getAnAdditionalChild = JSPaths::getAnAdditionalChild/2;
}

private module Resolver = PathResolver<ResolverConfig>;
