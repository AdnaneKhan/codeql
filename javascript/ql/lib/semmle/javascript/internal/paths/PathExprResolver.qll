private import javascript
private import semmle.javascript.internal.paths.PathResolver
private import semmle.javascript.internal.paths.TSConfig
private import semmle.javascript.internal.paths.PackageJsonEx

final private class FinalPathExpr = PathExpr;

private class RelevantPathExpr extends FinalPathExpr {
  RelevantPathExpr() { this = any(Import imprt).getImportedPath() }
}

/**
 * Gets the TSConfig file relevant for resolving `expr`.
 */
pragma[nomagic]
private TSConfig getTSConfigFromPathExpr(RelevantPathExpr expr) {
  result.getAnAffectedFile() = expr.getFile()
}

/**
 * Holds if `path` is relative, in the sense that it should be resolved relative to its enclosing folder.
 */
bindingset[path]
predicate isRelativePath(string path) { path.regexpMatch("\\.\\.?(?:[/\\\\].*)?") }

/**
 * Gets the NPM package name from the beginning of the given import path, e.g.
 * gets `foo` from `foo/bar`, and `@example/foo` from `@example/foo/bar`.
 */
pragma[nomagic]
private string getPackagePrefixFromPathExpr(RelevantPathExpr expr) {
  result = expr.getValue().regexpFind("^(@[^/\\\\]+[/\\\\])?[^@./\\\\][^/\\\\]*", _, _)
}

/**
 * Gets a folder name that is a common source folder name.
 */
string getASrcFolderName() { result = ["ts", "js", "src", "lib"] }

/**
 * Gets a folder name that is a common build output folder name.
 */
string getABuildOutputFolderName() { result = ["dist", "build", "out", "lib"] }

module JSPaths {
  /**
   * Gets an additional child of `base` to include when resolving JS paths.
   *
   * This accounts for two things:
   * - automatic file extensions (e.g `./foo` may resolve to `./foo.js`)
   * - mapping compiled-generated files back to their original source files
   */
  Container getAnAdditionalChild(Container base, string name) {
    // Automatically fill in file extensions
    result = base.(Folder).getJavaScriptFile(name)
    or
    // When importing a .js file, map to the original file that compiles to the .js file.
    exists(string stem |
      result = base.(Folder).getJavaScriptFile(stem) and
      name = stem + ".js"
    )
    or
    // Redirect './bar' to 'foo' given a tsconfig like:
    //
    //   { include: ["foo"], compilerOptions: { outDir: "./bar" }}
    //
    exists(TSConfig tsconfig |
      name =
        tsconfig.getCompilerOptions().getPropStringValue("outDir").regexpReplaceAll("^\\./", "") and
      base = tsconfig.getFolder() and
      result = tsconfig.getAnIncludedBaseContainer()
    )
    or
    // Heuristic version of the above based on commonly used source and build folder names
    exists(Folder folder | base = folder |
      folder = any(PackageJson pkg).getFolder() and
      name = getABuildOutputFolderName() and
      not exists(folder.getChildContainer(name)) and
      (
        result = folder.getChildContainer(getASrcFolderName())
        or
        result = folder
      )
    )
  }
}

/**
 * Gets an access to `__dirname`.
 */
private VarAccess dirname() { result.getName() = "__dirname" }

/** Holds if `add` is a relevant path expression of form `__dirname + expr`. */
private predicate prefixedByDirname(PathExpr expr) {
  expr = dirname()
  or
  prefixedByDirname(expr.(AddExpr).getLeftOperand())
  or
  prefixedByDirname(expr.(CallExpr).getArgument(0))
}

/**
 * Holds if `expr` matches a path mapping, and should thus be resolved as `newPath` relative to `base`.
 */
pragma[nomagic]
private predicate resolveViaPathMapping(RelevantPathExpr expr, Container base, string newPath) {
  // Handle tsconfig mappings such as `{ "paths": { "@/*": "./src/*" }}`
  exists(TSConfig config, string value |
    config = getTSConfigFromPathExpr(expr).getExtendedTSConfig*() and
    value = expr.getValue()
  |
    config.hasExactPathMappingTo(value, base) and
    newPath = ""
    or
    exists(string pattern |
      config.hasPrefixPathMappingTo(pattern, base) and
      value = pattern + newPath
    )
  )
  or
  // Handle imports referring to a package by name, where we have a package.json
  // file for that package in the codebase.
  //
  // This part only handles the "exports" property of package.json. "main" and "modules" are
  // handled further down because their semantics are easier to handle there.
  exists(PackageJsonEx pkg, string packageName, string remainder |
    packageName = getPackagePrefixFromPathExpr(expr) and
    pkg.getDeclaredPackageName() = packageName and
    remainder = expr.getValue().suffix(packageName.length()).regexpReplaceAll("^[/\\\\]", "")
  |
    // "exports": { ".": "./foo.js" }
    // "exports": { "./foo.js": "./foo/impl.js" }
    pkg.hasExactPathMappingTo(remainder, base) and
    newPath = ""
    or
    // "exports": { "./*": "./foo/*" }
    exists(string prefix |
      pkg.hasPrefixPathMappingTo(prefix, base) and
      remainder = prefix + newPath
    )
  )
}

/**
 * Holds if `expr` should be resolved as `path` relative to `base`.
 */
pragma[nomagic]
private predicate shouldResolve(RelevantPathExpr expr, Container base, string path) {
  // Relative paths are resolved from their enclosing folder
  path = expr.getValue() and
  isRelativePath(path) and
  base = expr.getFile().getParentContainer()
  or
  // Paths prefixed by __dirname should be resolved from the root dir, because __dirname
  // currently has a getValue() that returns its absolute path.
  prefixedByDirname(expr) and
  not exists(base.getParentContainer()) and // get root dir
  path = expr.getValue()
  or
  resolveViaPathMapping(expr, base, path)
  or
  // Resolve from baseUrl of relevant tsconfig.json file
  path = expr.getValue() and
  not isRelativePath(path) and
  base = getTSConfigFromPathExpr(expr).getBaseUrlFolder()
  or
  // If the path starts with the name of a package, but did not match any path mapping,
  // resolve relative to the enclosing directory of that package.
  // Note that `getFileFromFolderImport` may subsequently redirect this to the package's "main",
  // so we don't have to deal with that here.
  exists(PackageJson pkg, string packageName |
    packageName = getPackagePrefixFromPathExpr(expr) and
    pkg.getDeclaredPackageName() = packageName and
    path = expr.getValue().suffix(packageName.length()).regexpReplaceAll("^[/\\\\]", "") and
    base = pkg.getJsonFile().getParentContainer()
  )
}

private module ResolverConfig implements PathResolverSig {
  predicate shouldResolve(Container base, string path) { shouldResolve(_, base, path) }

  predicate getAnAdditionalChild = JSPaths::getAnAdditionalChild/2;
}

private module Resolver = PathResolver<ResolverConfig>;

private Container resolvePathExpr1(RelevantPathExpr expr) {
  exists(Container base, string path |
    shouldResolve(expr, base, path) and
    result = Resolver::resolve(base, path)
  )
}

private File guessPackageJsonMain1(PackageJsonEx pkg) {
  not exists(pkg.getMainFile()) and
  exists(Folder folder, Folder subfolder |
    folder = pkg.getFolder() and
    (
      subfolder = folder or
      subfolder = folder.getChildContainer(getASrcFolderName()) or
      subfolder =
        folder
            .getChildContainer(getASrcFolderName())
            .(Folder)
            .getChildContainer(getASrcFolderName())
    ) and
    result = subfolder.getJavaScriptFile("index")
  )
}

private File guessPackageJsonMain2(PackageJsonEx pkg) {
  not exists(pkg.getMainFile()) and
  not exists(guessPackageJsonMain1(pkg)) and
  result = pkg.getAFileInFilesArray()
}

private File getFileFromFolderImport(Folder folder) {
  result = folder.getJavaScriptFile("index")
  or
  // Note that unlike "exports" paths, "main" and "module" also take effect when the package
  // is imported via a relative path, e.g. `require("..")` targeting a folder with a package.json file.
  exists(PackageJsonEx pkg | pkg.getFolder() = folder |
    result = pkg.getMainFile()
    or
    result = guessPackageJsonMain1(pkg)
    or
    result = guessPackageJsonMain2(pkg)
  )
}

File resolvePathExpr(PathExpr expr) {
  result = resolvePathExpr1(expr)
  or
  result = getFileFromFolderImport(resolvePathExpr1(expr))
}

module Debug {
  class PathExprToDebug extends RelevantPathExpr {
    PathExprToDebug() { this.getValue() = "ai" }
  }

  query PathExprToDebug pathExprs() { any() }

  query TSConfig getTSConfigFromPathExpr_(PathExprToDebug expr) {
    result = getTSConfigFromPathExpr(expr)
  }

  query string getPackagePrefixFromPathExpr_(PathExprToDebug expr) {
    result = getPackagePrefixFromPathExpr(expr)
  }

  query predicate resolveViaPathMapping_(PathExprToDebug expr, Container base, string newPath) {
    resolveViaPathMapping(expr, base, newPath)
  }

  query predicate shouldResolve_(PathExprToDebug expr, Container base, string newPath) {
    shouldResolve(expr, base, newPath)
  }

  query Container resolvePathExpr1_(PathExprToDebug expr) { result = resolvePathExpr1(expr) }

  query File resolvePathExpr_(PathExprToDebug expr) { result = resolvePathExpr(expr) }

  // Some predicates that are usually small enough that they don't need restriction
  query File getPackageMainFile(PackageJsonEx pkg) { result = pkg.getMainFile() }

  query predicate guessPackageJsonMain1_ = guessPackageJsonMain1/1;

  query predicate guessPackageJsonMain2_ = guessPackageJsonMain2/1;

  query predicate getFileFromFolderImport_ = getFileFromFolderImport/1;
}
