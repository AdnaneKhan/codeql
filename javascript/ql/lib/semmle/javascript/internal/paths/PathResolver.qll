private import javascript

/**
 * Signature for modules to pass to `PathResolver`.
 */
signature module PathResolverSig {
  /**
   * Holds if `path` should be resolved to a file or folder, relative to `base`.
   */
  predicate shouldResolve(Container base, string path);

  /**
   * Gets an additional file or folder to consider a child of `base`.
   */
  default Container getAnAdditionalChild(Container base, string name) { none() }

  /**
   * Holds if `component` may be treated as `.` if it does not match a child.
   */
  default predicate isOptionalPathComponent(string component) { none() }

  /**
   * Holds if globs should be interpreted in the paths being resolved.
   *
   * The following types of globs are supported:
   * - `*` (matches any child)
   * - `**` (matches any child recursively)
   * - Complex patterns like `foo-*.txt` are also supported
   */
  default predicate allowGlobs() { none() }
}

/**
 * Provides a mechanism for resolving file paths relative to a given directory.
 */
module PathResolver<PathResolverSig Config> {
  private import Config

  private string getPathSegment(string path, int n) {
    shouldResolve(_, path) and
    result = path.replaceAll("\\", "/").splitAt("/", n)
  }

  pragma[nomagic]
  private string getPathSegmentAsGlobRegexp(string segment) {
    allowGlobs() and
    segment = getPathSegment(_, _) and
    segment.matches("%*%") and
    not segment = ["*", "**"] and // these are special-cased
    result = segment.regexpReplaceAll("[^a-zA-Z0-9*]", "\\\\$0").replaceAll("*", ".*")
  }

  private int getNumPathSegment(string path) {
    result = strictcount(int n | exists(getPathSegment(path, n)))
  }

  private Container getChild(Container base, string name) {
    result = base.(Folder).getChildContainer(name)
    or
    result = getAnAdditionalChild(base, name)
  }

  private Container resolve(Container base, string path, int n) {
    shouldResolve(base, path) and n = 0 and result = base
    or
    exists(Container current, string segment |
      current = resolve(base, path, n - 1) and
      segment = getPathSegment(path, n - 1)
    |
      result = getChild(current, segment)
      or
      segment = [".", ""] and
      result = current
      or
      segment = ".." and
      result = current.getParentContainer()
      or
      isOptionalPathComponent(segment) and
      not exists(getChild(current, segment)) and
      result = current
      or
      allowGlobs() and
      (
        segment = "*" and
        result = getChild(current, _)
        or
        segment = "**" and // allow empty match
        result = current
        or
        exists(string name |
          result = getChild(current, name) and
          name.regexpMatch(getPathSegmentAsGlobRegexp(segment))
        )
      )
    )
    or
    exists(Container current, string segment |
      current = resolve(base, path, n) and
      segment = getPathSegment(path, n)
    |
      // Follow child without advancing 'n'
      allowGlobs() and
      segment = "**" and
      result = getChild(current, _)
    )
  }

  /**
   * Gets the file or folder that `path` resolves to when resolved from `base`.
   *
   * Only has results for the `(base, path)` pairs provided by `shouldResolve`
   * in the instantiation of this module.
   */
  Container resolve(Container base, string path) {
    result = resolve(base, path, getNumPathSegment(path))
  }
}
