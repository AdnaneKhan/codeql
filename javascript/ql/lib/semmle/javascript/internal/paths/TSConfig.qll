private import javascript
private import semmle.javascript.internal.paths.PathExprResolver
private import semmle.javascript.internal.paths.PathResolver

class TSConfig extends JsonObject {
  TSConfig() {
    this.getJsonFile().getBaseName().matches("%tsconfig%.json") and
    this.isTopLevel()
  }

  /** Gets the folder containing this file. */
  Folder getFolder() { result = this.getJsonFile().getParentContainer() }

  /** Gets the `compilerOptions` object. */
  JsonObject getCompilerOptions() { result = this.getPropValue("compilerOptions") }

  /** Gets the string value in the `extends` property. */
  string getExtendsPath() { result = this.getPropStringValue("extends") }

  /** Gets the file referred to by the `extends` property. */
  File getExtendedFile() { result = Resolver::resolve(this.getFolder(), this.getExtendsPath()) }

  /** Gets the `TSConfig` file referred to by the `extends` property. */
  TSConfig getExtendedTSConfig() { result.getJsonFile() = this.getExtendedFile() }

  /** Gets the string value in the `baseUrl` property. */
  string getBaseUrlPath() { result = this.getCompilerOptions().getPropStringValue("baseUrl") }

  /** Gets the folder referred to by the `baseUrl` property in this file, not taking `extends` into account. */
  Folder getOwnBaseUrlFolder() {
    result = Resolver::resolve(this.getFolder(), this.getBaseUrlPath())
  }

  /** Gets the effective baseUrl folder for this tsconfig file. */
  Folder getBaseUrlFolder() {
    result = this.getOwnBaseUrlFolder()
    or
    not exists(this.getOwnBaseUrlFolder()) and
    result = this.getExtendedTSConfig().getBaseUrlFolder()
  }

  /** Gets a path mentioned in the `include` property. */
  string getAnIncludePath() {
    result = this.getPropStringValue("include")
    or
    result = this.getPropValue("include").(JsonArray).getElementStringValue(_)
  }

  /**
   * Gets a file or folder refenced by a path the `include` property, possibly
   * inherited from an extended tsconfig file.
   *
   * Does not include all the files within includes directories, use `getAnIncludedContainer` for that.
   */
  Container getAnIncludePathTarget() {
    result = Resolver::resolve(this.getFolder(), this.getAnIncludePath())
    or
    not exists(this.getPropValue("include")) and
    result = this.getExtendedTSConfig().getAnIncludePathTarget()
  }

  /**
   * Gets a file or folder inside the directory tree mentioned in the `include` property.
   */
  Container getAnIncludedContainer() {
    result = this.getAnIncludePathTarget()
    or
    result = this.getAnIncludedContainer().getAChildContainer()
  }

  private JsonObject getPathMappings() { result = this.getCompilerOptions().getPropValue("paths") }

  predicate hasPathMapping(string pattern, string newPath) {
    this.getPathMappings().getPropStringValue(pattern) = newPath
    or
    this.getPathMappings().getPropValue(pattern).(JsonArray).getElementStringValue(_) = newPath
  }

  predicate hasExactPathMapping(string pattern, string newPath) {
    this.hasPathMapping(pattern, newPath) and
    not pattern.matches("%*%")
  }

  predicate hasPrefixPathMapping(string pattern, string newPath) {
    this.hasPathMapping(pattern + "*", newPath + "*")
  }

  predicate hasExactPathMappingTo(string pattern, Container target) {
    exists(string newPath |
      this.hasExactPathMapping(pattern, newPath) and
      target = resolvePathMapping(this, newPath)
    )
    or
    this.getExtendedTSConfig().hasExactPathMappingTo(pattern, target)
  }

  predicate hasPrefixPathMappingTo(string pattern, Container target) {
    exists(string newPath |
      this.hasPrefixPathMapping(pattern, newPath) and
      target = resolvePathMapping(this, newPath)
    )
    or
    this.getExtendedTSConfig().hasPrefixPathMappingTo(pattern, target)
  }
}

/** For resolving paths in a tsconfig file, except `paths` mappings. */
private module ResolverConfig implements PathResolverSig {
  predicate shouldResolve(Container base, string path) {
    exists(TSConfig cfg |
      base = cfg.getFolder() and
      path = [cfg.getExtendsPath(), cfg.getBaseUrlPath(), cfg.getAnIncludePath()]
    )
  }

  predicate allowGlobs() { any() }
}

private module Resolver = PathResolver<ResolverConfig>;

/** For resolving `paths` mappings, since these require the baseURL to be resolved first. */
private module PathMappingResolverConfig implements PathResolverSig {
  additional predicate shouldResolve(TSConfig cfg, Container base, string path) {
    (cfg.hasExactPathMapping(_, path) or cfg.hasPrefixPathMapping(_, path)) and
    if isRelativePath(path)
    then base = cfg.getFolder() // relative paths are resolved relative to tsconfig.json
    else base = cfg.getBaseUrlFolder() // non-relative paths are resolved relative to the baseUrl
  }

  predicate shouldResolve(Container base, string path) { shouldResolve(_, base, path) }
}

private module PathMappingResolver = PathResolver<PathMappingResolverConfig>;

private Container resolvePathMapping(TSConfig cfg, string path) {
  exists(Container base |
    PathMappingResolverConfig::shouldResolve(cfg, base, path) and
    result = PathMappingResolver::resolve(base, path)
  )
}
