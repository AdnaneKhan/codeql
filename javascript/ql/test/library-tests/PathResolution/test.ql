import javascript
import semmle.javascript.internal.paths.PathExprResolver

query predicate importTarget(Import imprt, string value) {
  imprt.getImportedModule().getFile().getRelativePath() = value
}
