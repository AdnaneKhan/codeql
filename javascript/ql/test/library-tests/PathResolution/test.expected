| BaseUrl/src/main.ts:2:1:2:21 | import  ... /file"; | BaseUrl/lib/file.ts |
| BaseUrl/src/main.ts:3:1:3:24 | import  ... le.ts"; | BaseUrl/lib/file.ts |
| BaseUrl/src/main.ts:4:1:4:24 | import  ... le.js"; | BaseUrl/lib/file.ts |
| BaseUrl/src/main.ts:5:1:5:16 | import "../lib"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:6:1:6:22 | import  ... index"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:7:1:7:25 | import  ... ex.ts"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:8:1:8:25 | import  ... ex.js"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:11:1:11:18 | import "lib/file"; | BaseUrl/lib/file.ts |
| BaseUrl/src/main.ts:12:1:12:21 | import  ... le.ts"; | BaseUrl/lib/file.ts |
| BaseUrl/src/main.ts:13:1:13:21 | import  ... le.js"; | BaseUrl/lib/file.ts |
| BaseUrl/src/main.ts:14:1:14:13 | import "lib"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:15:1:15:19 | import "lib/index"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:16:1:16:22 | import  ... ex.ts"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:17:1:17:22 | import  ... ex.js"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:20:1:20:16 | import "@/file"; | BaseUrl/lib/file.ts |
| BaseUrl/src/main.ts:21:1:21:19 | import "@/file.ts"; | BaseUrl/lib/file.ts |
| BaseUrl/src/main.ts:22:1:22:19 | import "@/file.js"; | BaseUrl/lib/file.ts |
| BaseUrl/src/main.ts:23:1:23:11 | import "@"; | BaseUrl/lib/nostar.ts |
| BaseUrl/src/main.ts:24:1:24:17 | import "@/index"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:25:1:25:20 | import "@/index.ts"; | BaseUrl/lib/index.ts |
| BaseUrl/src/main.ts:26:1:26:20 | import "@/index.js"; | BaseUrl/lib/index.ts |
| Basic/Subdir/sub.ts:1:1:1:12 | import '..'; | Basic/index.ts |
| Basic/Subdir/sub.ts:2:1:2:13 | import '../'; | Basic/index.ts |
| Basic/Subdir/sub.ts:3:1:3:14 | import './..'; | Basic/index.ts |
| Basic/Subdir/sub.ts:4:1:4:15 | import './../'; | Basic/index.ts |
| Basic/index.ts:1:1:1:22 | import  ... r/sub'; | Basic/Subdir/sub.ts |
| DeclarationFiles/src/main.ts:1:1:1:22 | import  ... split"; | DeclarationFiles/lib/split.d.ts |
| DeclarationFiles/src/main.ts:1:1:1:22 | import  ... split"; | DeclarationFiles/lib/split.js |
| DeclarationFiles/src/main.ts:2:1:2:25 | import  ... it.js"; | DeclarationFiles/lib/split.d.ts |
| DeclarationFiles/src/main.ts:2:1:2:25 | import  ... it.js"; | DeclarationFiles/lib/split.js |
| DeclarationFiles/src/main.ts:3:1:3:27 | import  ... .d.ts"; | DeclarationFiles/lib/split.d.ts |
| DeclarationFiles/src/main.ts:5:1:5:27 | import  ... cript"; | DeclarationFiles/lib/typescript.ts |
| DeclarationFiles/src/main.ts:6:1:6:30 | import  ... pt.js"; | DeclarationFiles/lib/typescript.ts |
| DeclarationFiles/src/main.ts:7:1:7:32 | import  ... .d.ts"; | DeclarationFiles/lib/typescript.d.ts |
| DirnameImports/main.js:5:16:5:48 | require ... et.js') | DirnameImports/target.js |
| DirnameImports/main.js:8:22:8:63 | require ... t.js')) | DirnameImports/target.js |
| DirnameImports/main.js:11:16:11:55 | require ... et.js') | DirnameImports/nested/target.js |
| DirnameImports/main.js:14:16:14:60 | require ... es.ts') | import-packages.ts |
| DirnameImports/main.js:18:17:18:64 | require ... et.js') | DirnameImports/nested/target.js |
| DirnameImports/main.js:21:20:21:57 | require ... et.js') | DirnameImports/target.js |
| Extended/src/main.ts:2:1:2:21 | import  ... /file"; | Extended/lib/file.ts |
| Extended/src/main.ts:3:1:3:24 | import  ... le.ts"; | Extended/lib/file.ts |
| Extended/src/main.ts:4:1:4:24 | import  ... le.js"; | Extended/lib/file.ts |
| Extended/src/main.ts:5:1:5:16 | import "../lib"; | Extended/lib/index.ts |
| Extended/src/main.ts:6:1:6:22 | import  ... index"; | Extended/lib/index.ts |
| Extended/src/main.ts:7:1:7:25 | import  ... ex.ts"; | Extended/lib/index.ts |
| Extended/src/main.ts:8:1:8:25 | import  ... ex.js"; | Extended/lib/index.ts |
| Extended/src/main.ts:11:1:11:18 | import "lib/file"; | Extended/lib/file.ts |
| Extended/src/main.ts:12:1:12:21 | import  ... le.ts"; | Extended/lib/file.ts |
| Extended/src/main.ts:13:1:13:21 | import  ... le.js"; | Extended/lib/file.ts |
| Extended/src/main.ts:14:1:14:13 | import "lib"; | Extended/lib/index.ts |
| Extended/src/main.ts:15:1:15:19 | import "lib/index"; | Extended/lib/index.ts |
| Extended/src/main.ts:16:1:16:22 | import  ... ex.ts"; | Extended/lib/index.ts |
| Extended/src/main.ts:17:1:17:22 | import  ... ex.js"; | Extended/lib/index.ts |
| Extended/src/main.ts:20:1:20:16 | import "@/file"; | Extended/lib/file.ts |
| Extended/src/main.ts:21:1:21:19 | import "@/file.ts"; | Extended/lib/file.ts |
| Extended/src/main.ts:22:1:22:19 | import "@/file.js"; | Extended/lib/file.ts |
| Extended/src/main.ts:23:1:23:11 | import "@"; | Extended/lib/nostar.ts |
| Extended/src/main.ts:24:1:24:17 | import "@/index"; | Extended/lib/index.ts |
| Extended/src/main.ts:25:1:25:20 | import "@/index.ts"; | Extended/lib/index.ts |
| Extended/src/main.ts:26:1:26:20 | import "@/index.js"; | Extended/lib/index.ts |
| NoBaseUrl/src/main.ts:2:1:2:21 | import  ... /file"; | NoBaseUrl/lib/file.ts |
| NoBaseUrl/src/main.ts:3:1:3:24 | import  ... le.ts"; | NoBaseUrl/lib/file.ts |
| NoBaseUrl/src/main.ts:4:1:4:24 | import  ... le.js"; | NoBaseUrl/lib/file.ts |
| NoBaseUrl/src/main.ts:5:1:5:16 | import "../lib"; | NoBaseUrl/lib/index.ts |
| NoBaseUrl/src/main.ts:6:1:6:22 | import  ... index"; | NoBaseUrl/lib/index.ts |
| NoBaseUrl/src/main.ts:7:1:7:25 | import  ... ex.ts"; | NoBaseUrl/lib/index.ts |
| NoBaseUrl/src/main.ts:8:1:8:25 | import  ... ex.js"; | NoBaseUrl/lib/index.ts |
| NoBaseUrl/src/main.ts:20:1:20:16 | import "@/file"; | NoBaseUrl/lib/file.ts |
| NoBaseUrl/src/main.ts:21:1:21:19 | import "@/file.ts"; | NoBaseUrl/lib/file.ts |
| NoBaseUrl/src/main.ts:22:1:22:19 | import "@/file.js"; | NoBaseUrl/lib/file.ts |
| NoBaseUrl/src/main.ts:23:1:23:11 | import "@"; | NoBaseUrl/lib/nostar.ts |
| NoBaseUrl/src/main.ts:24:1:24:17 | import "@/index"; | NoBaseUrl/lib/index.ts |
| NoBaseUrl/src/main.ts:25:1:25:20 | import "@/index.ts"; | NoBaseUrl/lib/index.ts |
| NoBaseUrl/src/main.ts:26:1:26:20 | import "@/index.js"; | NoBaseUrl/lib/index.ts |
| import-packages.ts:1:1:1:32 | import  ... /main'; | PackageWithMain/main.js |
| import-packages.ts:2:1:2:36 | import  ... -main'; | PackageWithMain/main.js |
| import-packages.ts:4:1:4:33 | import  ... eMain'; | PackageWithModuleMain/main.js |
| import-packages.ts:5:1:5:43 | import  ... -main'; | PackageWithModuleMain/main.js |
| import-packages.ts:10:1:10:39 | import  ... ports'; | PackageWithExports/main.js |
| import-packages.ts:11:1:11:49 | import  ... -file'; | PackageWithExports/fake-file-impl.js |
| import-packages.ts:12:1:12:48 | import  ... r/foo'; | PackageWithExports/star-impl/foo.js |
| import-packages.ts:14:1:14:28 | import  ... xFile'; | PackageIndexFile/index.js |
| import-packages.ts:15:1:15:42 | import  ... -file'; | PackageIndexFile/index.js |
| import-packages.ts:17:1:17:25 | import  ... uess1'; | PackageGuess1/src/index.ts |
| import-packages.ts:18:1:18:33 | import  ... uess1'; | PackageGuess1/src/index.ts |
| import-packages.ts:20:1:20:25 | import  ... uess2'; | PackageGuess2/blah/stuff.ts |
| import-packages.ts:21:1:21:33 | import  ... uess2'; | PackageGuess2/blah/stuff.ts |
| import-packages.ts:23:1:23:30 | import  ... inExt'; | PackageWithMainExt/lib/main.ts |
| import-packages.ts:24:1:24:40 | import  ... n-ext'; | PackageWithMainExt/lib/main.ts |
| import-packages.ts:26:1:26:46 | import  ... oo.js'; | TSConfigOutDir/src/foo.ts |
| import-packages.ts:28:1:28:24 | import  ... older'; | MainIsFolder/src/index.ts |
| import-packages.ts:29:1:29:33 | import  ... older'; | MainIsFolder/src/index.ts |
| import-packages.ts:31:1:31:27 | import  ... nsSrc'; | DistContainsSrc/src/foo.ts |
